Purecap ddb

bsdjhb opened the pull request:
These are MI changes for DDB support.  Mostly it consists of helpers for DDB commands that accept user-supplied addresses.  This takes a conservative approach and does not try to bound symbol addresses based on the symbol information.

jrtc27 commented on sys/ddb/db_main.c:
```diff
@@ -191,7 +202,7 @@ X_db_symbol_values(db_symtab_t *symtab, c_db_sym_t sym, const char **namep,
 
 int
-db_fetch_ksymtab(vm_offset_t ksym_start, vm_offset_t ksym_end,
+db_fetch_ksymtab(vm_pointer_t ksym_start, vm_pointer_t ksym_end,
```
ksym_end can remain a virtual address

bsdjhb replied:
While it can, it's a bit odd of an interface to pass around a pointer in one case and a VA in the other, and most of the other similar APIs in the kernel linker and other places were simpler to leave as pairs.

bsdjhb replied:
The other open question is that ksym_start and ksym_end come from the metadata from the loader, and I think we aren't sure yet what the loader will pass: will it pass cap pointers to the kernel (it probably should?) or it will it pass virtual addresses.

jrtc27 replied:
I imagine it should pass whatever the native ABI of the loaded kernel is.

jrtc27 commented on sys/ddb/ddb.h:
```diff
@@ -83,7 +83,9 @@ int	DB_CALL(db_expr_t, db_expr_t *, int, db_expr_t[]);
  */
-extern vm_offset_t ksymtab, kstrtab, ksymtab_size, ksymtab_relbase;
+extern vm_pointer_t ksymtab, kstrtab;
+extern vm_offset_t ksymtab_relbase;
+extern vm_size_t ksymtab_size;
```
```suggestion
extern vm_size_t ksymtab_size;
extern vm_offset_t ksymtab_relbase;
```
keep them in the same order?

bsdjhb replied:
Done

jrtc27 commented:
Hm, losing all the types is a little sad even if they're not really doing anything. Maybe we should be doing something that's a blend of __DECONST and __USER_CAP_OBJ where we provide the type and it does the sizeof and cast for us?

jrtc27 commented:
I also note you're inconsistent about whether there's a blank line before the CHERI CHANGES blocks.

bsdjhb commented on sys/geom/geom_subr.c:
```diff
@@ -1574,18 +1574,22 @@ DB_SHOW_COMMAND(geom, db_show_geom)
 		}
 	} else {
-		switch (g_valid_obj((void *)addr)) {
+		switch (g_valid_obj((void *)(uintptr_t)addr)) {
```
Eh, this probably doesn't work as-is.

bsdjhb replied:
Oh, it sort of does, but g_valid_obj itself needs some changes to only compare addresses.

bsdjhb commented on sys/geom/geom_subr.c:
```diff
@@ -1641,7 +1645,7 @@ DB_SHOW_COMMAND(bio, db_show_bio)
 
 	if (have_addr) {
-		bp = (struct bio *)addr;
+		bp = DB_DATA_PTR(addr, sizeof(*bp));
```
We could perhaps have a `DB_DATA_PTR_OBJ` that does `sizeof(*arg)`.  Many uses of `DB_DATA_PTR` don't have a typed variable.

bsdjhb commented:
For the blank lines, originally CHERI CHANGES are right up against the bottom of the file, but I find that 1) that makes it a bit more of a PITA for merge conflicts, and 2) it is harder to read (and doesn't match the style where there's always a blank line after the license comment at the start of a file).  For those reasons I've switched to adding a blank line for new CHERI CHANGES blocks that I've added in the purecap branch, but I haven't changed existing blocks in the branch.  We could either never use the blank line, or use blank lines for new ones as they get merged from purecap into dev perhaps?  I'll defer to @brooksdavis.

bsdjhb commented:
Hmmm, the problem with something like `DB_DATA_PTR_OBJ` is that you would need to use the variable twice unless the assignment is in the macro itself which would be too gross, that is:

```
     struct bio *bp;

     bp = DB_DATA_PTR_OBJ(addr, bp);
```

Instead of

```
     bp = DB_DATA_PTR(addr, sizeof(*bp));
```

The problem here is that `addr` is never typed unlike the uses of `__USER_CAP_OBJ` where the argument is typed.  I'm not sure that really buys you much since you still need the second argument to the macro.

I guess you are asking more for something like `DB_DATA_PTR(addr, struct bio)`.  Perhaps that is a bit shorter as it removes the need for the explicit `sizeof()` around the type.  It does mean you can't use the `sizeof(*bp)` pattern, though you'd get a compile error if you get the type wrong in those cases.

jrtc27 commented:
> I guess you are asking more for something like `DB_DATA_PTR(addr, struct bio)`. Perhaps that is a bit shorter as it removes the need for the explicit `sizeof()` around the type. It does mean you can't use the `sizeof(*bp)` pattern, though you'd get a compile error if you get the type wrong in those cases.

Yes, that's what I meant, precisely due to the issues you mentioned (well, the second one, the first option only works if you're assigning to a variable and even then is rather gross). That's what I was trying to communicate by bringing up `__DECONST`.

jrtc27 commented:
(and the duplicate types already exist as casts upstream, so losing `sizeof(*bp)` etc is no worse than upstream)

bsdjhb commented:
Ok, I've reworked to pass types in.  For now it's a separate fixup commit, but if this works better I'll squash it into the second commit.  I also threw in mips/mips/db_interface.c fixes mirroring those for riscv that are in the second commit.

jrtc27 commented on sys/kern/subr_vmem.c:
```diff
@@ -1702,7 +1702,7 @@ DB_SHOW_COMMAND(vmemdump, vmemdump)
 	}
 
-	vmem_dump((const vmem_t *)addr, db_printf);
+	vmem_dump(DB_DATA_PTR(addr, vmem_t), db_printf);
```
const? (and below)

bsdjhb replied:
Oh, yes, I switched to keeping the const part way in and forgot to go back and fix the few from before.

jrtc27 commented:
Yeah I think this looks good.

brooksdavis commented:
> For the blank lines, originally CHERI CHANGES are right up against the bottom of the file, but I find that 1) that makes it a bit more of a PITA for merge conflicts, and 2) it is harder to read (and doesn't match the style where there's always a blank line after the license comment at the start of a file). For those reasons I've switched to adding a blank line for new CHERI CHANGES blocks that I've added in the purecap branch, but I haven't changed existing blocks in the branch. We could either never use the blank line, or use blank lines for new ones as they get merged from purecap into dev perhaps? I'll defer to @brooksdavis.

My motivation for not adding blank lines was that the blocks can then be mechanically removed leaving no diffs to upstream. Having blank lines and trimming them would also have that effect, but require a slightly more complicated awk script when the time comes. (Likewise the switch to `//` comments in kernel.)

bsdjhb commented:
Restored const and removed the blank lines.
